sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java

```diff
@@ -0,0 +1,275 @@
+        this.maxVarArgOccurrences = -1;
+    }
+
+    public static class Builder {
+
+        private FuncParams funcParams;
+
+        public Builder(Param... parameters) {
+            this.funcParams = new FuncParams(parameters);
+        }
+
+        public Builder(List<DataType> fixedSignature) {
+            Param[] params = new Param[fixedSignature.size()];
+            for (int i = 0; i < params.length; i++) {
+                params[i] = Param.of(fixedSignature.get(i));
+            }
+            this.funcParams = new FuncParams(params);
+        }
+
+        /**
+         * Adds an optional and variable number of occurrences of the
+         * following parameters.
+         * @param parameters The types used in the var arg parameters.
+         * @return FuncParams
+         */
+        public Builder withVarArgs(Param... parameters) {
+            this.funcParams.varArgParameters = parameters;
+            return this;
+        }
+
+        /**
+         * Adds an optional and a fixed upper number of occurrences of the
+         * following parameters.
+         * @param parameters The types used in the var arg parameters.
+         * @return FuncParams
+         */
+        public Builder withVarArgs(int limitOfOccurrences, Param... parameters) {
+            Preconditions.checkArgument(limitOfOccurrences > 0,
+                "The minimum limit for the number of occurences is 1");
+            this.funcParams.varArgParameters = parameters;
+            this.funcParams.maxVarArgOccurrences = limitOfOccurrences;
+            return this;
+        }
+
+        /**
+         * One or multiple var args which operate completely independently of each other.
+         * @param parameters The parameters which are independent of each other.
+         * @return A Builder which can be used
+         */
+        public Builder withIndependentVarArgs(Param... parameters) {
+            withVarArgs(parameters);
+            this.funcParams.varArgsIndependent = true;
+            return this;
+        }
+
+        public FuncParams build() {
+            try {
+                return this.funcParams;
+            } finally {
+                // prevent reuse of the Builder
+                this.funcParams = null;
+            }
+        }
+    }
+
+    public static Builder of(Param... parameters) {
+        return new Builder(parameters);
+    }
+
+    public static Builder of(List<DataType> fixedSignature) {
+        Param[] params = new Param[fixedSignature.size()];
+        for (int i = 0; i < params.length; i++) {
+            params[i] = Param.of(fixedSignature.get(i));
+        }
+        return of(params);
+    }
+
+    private void resetBoundTypes() {
+        for (Param parameter : parameters) {
+            parameter.clear();
+        }
+        for (Param parameter : varArgParameters) {
+            parameter.clear();
+        }
+    }
+
+    private void bindParams(List<? extends FuncArg> params) {
```

Comment:
I guess this method could use some explanation

Reply:
My opinion is that without reading the method body, just by reading the signature/docs I should be able to guess what a method is doing - in this case (and `bind`) that's not the case.